### ISSUE TITLE
Improve HTTP POST compatibility with CURL

### ISSUE THREAD

**Issue by Akandesh** [uNetworking/uWebSockets#1023]
When sending 800 bytes it's fast. Localhost its 0ms. Sending 1280 bytes takes 10 seconds. Although on my VPS it only takes around 1550ms. Something is clearly messed up somewhere my side or yours. I've narrowed it down to the lib itself.

My test is simply
```
require('uWebSockets.js').App(
  // {

  //   /* There are tons of SSL options */
  //   key_file_name: 'private.pem',
  //   cert_file_name: 'public.pem',
    
  // }
  ).ws('/*', {
  
    /* For brevity we skip the other events */
    message: (ws, message, isBinary) => {
      let ok = ws.send(message, isBinary);
    }
    
  }).any('/*', (res, req) => {
  
    /* Let's deny all Http */
    console.log("Got hit");
    res.end('Nothing to see here!');
    
  }).listen(9001, (listenSocket) => {
  
    if (listenSocket) {
      console.log('Listening to port 9001');
    }
    
  });
```

My C++ client is using curl with these options:
```	
        headers = curl_slist_append( headers, ( "Content-Type: application/zip" ) );
        curl_easy_setopt( curl, CURLOPT_VERBOSE, 0L );
	curl_easy_setopt( curl, CURLOPT_HTTPHEADER, headers );
	curl_easy_setopt( curl, CURLOPT_URL, "http://127.0.0.1:9001/" );
	curl_easy_setopt( curl, CURLOPT_CUSTOMREQUEST, "POST" );
	curl_easy_setopt( curl, CURLOPT_POSTFIELDS, out_string.c_str( ) );
	curl_easy_setopt( curl, CURLOPT_POSTFIELDSIZE, out_string.size( ) );
	curl_easy_setopt( curl, CURLOPT_WRITEFUNCTION, &write_data );
	curl_easy_setopt( curl, CURLOPT_NOSIGNAL, 1 );
	curl_easy_setopt( curl, CURLOPT_POST, 1L );
	curl_easy_setopt( curl, CURLOPT_SSL_VERIFYPEER, 0L );
```

I just simply generate a string with 2000 characters then I compress them using zlib down to 1300ish then send.

To measure the performance I calculate the time curl_easy_perform takes.

**Comment by alexhultman** [comment]
Can you post your entire C++ code?

One potential bug is this:

1. Curl sends Content-Length of a few MB, we parse that and expect those bytes within 10 seconds or else we close the connection due to timeout.
2. Curl also sends the header "Expect: 100-continue", meaning it will only send that data if we send the "100 Continue", otherwise, if we send 200 OK the curl client will not actually send the data, even though it just declared its content-length to be multiple MB.
3. Curl is fine with the 200 OK, and considers the request responded to (should be very fast)
4. However µWS is still waiting for those MB of payload since it does not understand the 100 Continue thing, causing whatever you send to the kept-alive connection to just fail within a maximum of 10 seconds or until filled up.

Basically, the flaw you describe should only happen if you have some kind of loop, reusing the same connection many times. That's why I need to look at your entire C++ code because one single request, or multiple separate, cannot trigger this issue.

**Comment by alexhultman-2fa** [comment]
https://curl.haxx.se/mail/lib-2017-07/0013.html

That thread is highly relevant here. Curl changes behavior at 1kb (just about what you report) and will wait 1 second if there is no 100 Continue. The way you respond immediately with 200 OK will like mentioned above break the following requests.

> You can do any amount of calls to curl_easy_perform while using the same easy_handle. If you intend to transfer more than one file, you are even encouraged to do so. libcurl will then attempt to re-use the same connection for the following transfers, thus making the operations faster, less CPU intense and using less network resources. Just note that you will have to use curl_easy_setopt between the invokes to set options for the following curl_easy_perform.

Yeah I bet you have a loop reusing the same easy_handle. That would explain the 10 seconds, or possibly 1 second.

**Comment by Akandesh** [comment]
Yeah, I'm sending it in a while loop. I can upload the entire test project if you want but I don't think it's needed. The important stuff is

```cpp
	std::string long_string, small_string;
	init_string( long_string, 2000 );
	init_string( small_string, 500 );
	curl = curl_easy_init( );
	//printf( "Compressed: %i : %i\n", compressed_long.size( ), compressed_small.size( ) );
	while ( !GetAsyncKeyState( VK_ESCAPE ) ) {
		send_data( long_string, true, false );
		Sleep( 400 );
		send_data( small_string, true, false );
		Sleep( 1000 );
	}
	curl_easy_cleanup( curl );
	return 0;
```

**Comment by alexhultman-2fa** [comment]
Great, yeah it is the curl behavior then. Will fix.

**Comment by alexhultman** [comment]
It does respond correctly now. Thanks for reporting.
